Internal refactoring of Matrix Vector Multiplication routines

YvanMokwinski opened the pull request:
- removing deprecated function bsrmv_ex
- adding a flag fallback_algorithm for internal mv routines
- replacing mat_info with corresponding csrmv/bsrmv info for internal mv routines
- adding csrmv/bsrmv info in v2_spmv descriptor
- no_fallback algorithm for v2_spmv

YvanMokwinski committed:
- removing deprecated function bsrmv_ex
- adding a flag fallback_algorithm for internal mv routines
- replacing mat_info with corresponding csrmv/bsrmv info for internal mv routines
- adding csrmv/bsrmv info in v2_spmv descriptor
- no_fallback algorithm for v2_spmv

YvanMokwinski committed:
putting back rocsparse_cscmv_info pointer definition

YvanMokwinski committed:
removing a commented line of code

YvanMokwinski committed:
cleanup

YvanMokwinski committed:
updating the change log

ntrost57 commented on library/src/level2/rocsparse_coomv.cpp:
```diff
@@ -31,20 +31,21 @@
+                                        const void*,
+                                        const void*,
+                                        void*,
+                                        bool);
```
is it best practise like that? It has no impact right?

ntrost57 commented on CHANGELOG.md:
```diff
@@ -15,6 +15,8 @@ Documentation for rocSPARSE is available at
 * The deprecated `rocsparse_spmv_ex` routine
+* The deprecated `rocsparse_sbsrmv_ex`, `rocsparse_dbsrmv_ex`, `rocsparse_cbsrmv_ex` and `rocsparse_zbsrmv_ex`  routines
+* The deprecated `rocsparse_sbsrmv_ex_analysis`, `rocsparse_dbsrmv_ex_analysis`, `rocsparse_cbsrmv_ex_analysis` and `rocsparse_zbsrmv_ex_analysis`  routines
 
```
Maybe also add that we added additional warnings when falling back to other algorithms.

YvanMokwinski replied:
These warnings are only available when calling `rocsparse_debug_enable`, or exporting the environment variable ROCSPARSE_DEBUG with an assigned value of 1. Do you still want me to add the note?

YvanMokwinski replied:
I added the note.

amd-jnovotny commented on CHANGELOG.md:
```diff
@@ -15,6 +15,8 @@ Documentation for rocSPARSE is available at
 ### Removed
 
 * The deprecated `rocsparse_spmv_ex` routine
+* The deprecated `rocsparse_sbsrmv_ex`, `rocsparse_dbsrmv_ex`, `rocsparse_cbsrmv_ex` and `rocsparse_zbsrmv_ex`  routines
+* The deprecated `rocsparse_sbsrmv_ex_analysis`, `rocsparse_dbsrmv_ex_analysis`, `rocsparse_cbsrmv_ex_analysis` and `rocsparse_zbsrmv_ex_analysis`  routines
```
```suggestion
* The deprecated `rocsparse_sbsrmv_ex`, `rocsparse_dbsrmv_ex`, `rocsparse_cbsrmv_ex`, and `rocsparse_zbsrmv_ex`  routines
* The deprecated `rocsparse_sbsrmv_ex_analysis`, `rocsparse_dbsrmv_ex_analysis`, `rocsparse_cbsrmv_ex_analysis`, and `rocsparse_zbsrmv_ex_analysis`  routines
```

amd-jnovotny reviewed (approved):
Docs and changelog OK. One very minor suggestion.

YvanMokwinski committed:
Removing tests that calls algorithm with non-implemented configuration since v2_spmv does not allow fallbacks.

YvanMokwinski committed:
Updating changelog
Use rocsparse_spmv_alg_csr_rowsplit rather than the deprecated rocsparse_spmv_alg_csr_stream
Updating the documentation about the non-optional analysis

amd-jnovotny commented on CHANGELOG.md:
```diff
@@ -10,7 +10,7 @@ Documentation for rocSPARSE is available at
 ### Added
 * Adds `SpGEAM` generic routine for computing sparse matrix addition in CSR format
-* Adds `v2_SpMV` generic routine for computing sparse matrix vector multiplication
+* Adds `v2_SpMV` generic routine for computing sparse matrix vector multiplication. As opposed to the deprecated `rocsparse_spmv` routine, this routine does not use a fallback algorithm if a non-implemented configuration is encountered and will return an error in such a case. For the deprecated routine `rocsparse_spmv`, the user can enable warning messages in situations where a fallback algorithm ...
```
```suggestion
* Adds `v2_SpMV` generic routine for computing sparse matrix vector multiplication. As opposed to the deprecated `rocsparse_spmv` routine, this routine does not use a fallback algorithm if a non-implemented configuration is encountered and will return an error in such a case. For the deprecated routine `rocsparse_spmv`, the user can enable warning messages in situations where a fallback algorithm is used by either calling upfront the routine `rocsparse_enable_debug` or exporting the variable `ROCSPARSE_DEBUG` (with the shell command `export ROCSPARSE_DEBUG=1`).
```

amd-jnovotny reviewed (approved):
Just a minor formatting adjustment suggestion (code ticks)

YvanMokwinski committed:
Fixing a wrong condition

YvanMokwinski committed:
Update CHANGELOG.md

Co-authored-by: Jeffrey Novotny <jnovotny@amd.com>

YvanMokwinski committed:
Update CHANGELOG.md

Co-authored-by: Jeffrey Novotny <jnovotny@amd.com>

jsandham commented on library/src/level2/rocsparse_gebsrmv.cpp:
```diff
@@ -397,6 +397,9 @@ rocsparse_status rocsparse::gebsrmv_template(rocsparse_handle          handle, /
     if(row_block_dim == 1 && col_block_dim == 1)
     {
+        static constexpr bool fallback_algorithm = true;
+        static constexpr bool no_force_conj      = false;
```
Should this be:

static constexpr bool force_conj = false;

YvanMokwinski replied:
That was on purpose. I found it more explicit when passing the parameter to the routine, but okay if you request it.
